src/nvim/drawline.c

```diff
@@ -3036,8 +3037,8 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, s
       virt_line_index = -1;
       virt_line_flags = 0;
       // When the filler lines are actually below the last line of the
-      // file, don't draw the line itself, break here.
-      if (wlv.filler_todo == 0 && (wp->w_botfill || end_fill)) {
+      // file or line line is concealed, don't draw the line itself, break here.
+      if (wlv.filler_todo == 0 && (wp->w_botfill || end_fill || concealed)) {
```

Comment:
It seems that `concealed` actually needs to be checked at almost every other place where `end_fill` is checked, except for the one in the `if (statuscol.draw)` branch.